### ISSUE TITLE
refreshing issues

### ISSUE THREAD

**Issue by Xeverous** [cycfi/elements#142]
So I have started to work a bit with buttons and menus (I'm a bit unsatisfied with the API but that will be a separate issue) and noticed a common problem: refreshing

- Sometimes a button performs an action that modifies or adds other UI elements, and they do not render correctly when not refreshed, but the having to capture so many shared pointers quickly gets out of hand.
- Sometimes refreshing does not help. Refresh a label, it's new text is rendered but if the space occupied by the text is different (eg larger) it is rendered on top of surrounding elements, which still use its old size. As if I had to refresh few levels higher in the hierarchy - resizing the window fixes the problem (which AFAIK refreshes everything).

**Comment by djowel** [comment]
MVCE please. Without you providing an MVCE, I can never know, but I suspect that is just using the "outward" param of refresh, if not, it is always possible to render the whole view (link below).  For that matter, it's best to always explain a use-case, together with an MVCE.

https://github.com/cycfi/elements/blob/master/lib/include/elements/view.hpp#L52-L55

Before ranting about dissatisfaction, how about asking (nicely!) first if a certain use-case is possible, before assuming that it is not? Constant complaints and dissatisfaction gets tiring, you know, and does not have its place in a collaborative environment, especially one done freely. Keep in mind that I am doing this in my free time and I do not have the obligation to convince you, nor please you in anything. I would appreciate it if you change your tone, so instead of complaints and condescension, making it more like requests with a bit more humility instead.

**Comment by Xeverous** [comment]
MCVE (works with `vtile`, does not work with `htile`):

Note: my project has obviously much more complex UI and there are more complex cases. This is just one of them that I found easy to reproduce.

```cpp
#include <elements.hpp>

using namespace cycfi::elements;

auto make_button(std::string text, std::function<void(bool)> on_click)
{
     auto but = button(std::move(text));
     but.on_click = std::move(on_click);
     return but;
}

auto make_widgets(view& v)
{
   auto lb = share(label(""));

   auto on_click = [lb, &v](bool)
   {
      lb->set_text(lb->get_text() + std::string(" test"));
      v.refresh(*lb);
   };

   return htile(
      label("test"),
      hold(lb),
      make_button("change label", on_click)
   );
}

int main(int argc, char* argv[])
{
   app _app(argc, argv);
   window _win(_app.name());
   _win.on_close = [&_app]() { _app.stop(); };

   view view_(_win);

   view_.content(
      make_widgets(view_),
      box(rgba(35, 35, 37, 255))
   );

   _app.run();
   return 0;
}
```

> Before ranting about dissatisfaction, how about asking (nicely!) first if a certain use-case is possible, before assuming that it is not? Constant complaints and dissatisfaction gets tiring, you know, and does not have its place in a collaborative environment, especially one done freely.

I think there is some unwanted difference / misunderstanding between what I write and what you perceive (remember the sarcasm?). By no means I wanted such tone, perhaps it's cultural difference.

By "I'm a bit unsatisfied with the API but that will be a separate issue" I meant:

- I will describe my use case in a separate issue and why/why not I like/dislike certain interface
- I will try PRs when agreed upon an interface with you

and btw, I'm waiting on the cairo leaks PR for your response

**Comment by djowel** [comment]
> Note: my project has obviously much more complex UI and there are more complex cases. This is just one of them that I found easy to reproduce.

Have you tried providing the `outward` param or just refreshing the whole view (no params)?

> I think there is some unwanted difference / misunderstanding between what I write and what you perceive (remember the sarcasm?). By no means I wanted such tone, perhaps it's cultural difference.

Perhaps... I suggest a less demanding tone. For example:

> and btw, I'm waiting on the cairo leaks PR for your response

Instead of that, how about: "You must be busy, but if you have time, could you please have a look at the cairo leaks PR?"? Which transforms the demand into an unassuming request. Kinder words such as "please" matters in a text-only communication medium. The former sounds as if you are my boss and that I am obliged to satisfy your demand :-).

**Comment by Xeverous** [comment]
> Have you tried providing the outward param or just refreshing the whole view (no params)?

After changing an element, I actually don't know how to refresh properly.

- Should I call `view.refresh(changed_element)`? Currently, it does not always work.
- Should I call `view.refresh(changed_element, outward)`? How would I get the right `outward` value?
- Should I call `view.refresh()` to refresh the entire view?
- Something else?

___

I guess you are very busy then? You wrote earlier that you are not ignoring me; I was just worried because I wrote a simple yes/no question and did not get a reponse within 1 day so I thought you might lose/miss the thread/mail about it (whatever you use to track topics, I'm marking mails from GitHub as unread).

**Comment by djowel** [comment]
I'll have a look at your use case ASAP.

> * Should I call `view.refresh(changed_element)`? Currently, it does not always work.
> * Should I call `view.refresh(changed_element, outward)`? How would I get the right `outward` value?

The only reason why it won't work is if it is inside a proxy that does some decorations, alignment, etc. around the element. Typically, you know how this is arranged. For example, if you have an element inside a margin, inside an align, then you'd want to refresh the outermost element that encloses the element in question. In this case 3 or 4 to be safe.

> * Should I call `view.refresh()` to refresh the entire view?

`view.refresh(element)` is a conservative (draw only what's needed) refresh. In many cases, esp. in a busy interactive UI, you can simply call `view.refresh()` to redraw the entire thing. That's fine. Some GUIs, e.g. the "immediate mode" UIs **always** redraws the entire scene every time!

> I guess you are very busy then? You wrote earlier that you are not ignoring me; I was just worried because I wrote a simple yes/no question and did not get a reponse within 1 day so I thought you might lose/miss the thread/mail about it (whatever you use to track topics, I'm marking mails from GitHub as unread).

Yeah, we're about to release some hardware. That said, Elements is VERY IMPORTANT to me and I can promise you that no "issue" will be ignored. I am simply reallocating time as best as I can.

**Comment by Xeverous** [comment]
> `view.refresh(element)` is a conservative (draw only what's needed) refresh. In many cases, esp. in a busy interactive UI, you can simply call `view.refresh()` to redraw the entire thing.

I'm surprised you managed to implement lazy-evaluated redrawing. Initially I thought the library is redrawing the screen every frame. I'm even more surprised how little actual state the `view` class holds to note what is dirty.

I will try refreshing the entire view and see if it works. Granular refreshing seems to be a premature optimization for my project and it requires a lot of extra code and work to get right. And there is always a chance for a missing-refreshing-bug when I change some hierarchy but don't update some refresh code.

**Comment by djowel** [comment]
> Do you know any inconvenient about avoiding threading problems by this method?

This method is perfectly thread safe. ASIO takes care of the threading issues. View invalidation is guaranteed to happen only in the main thread.

**Comment by Xeverous** [comment]
`view.refresh()` for sure has some bugs OR I should use a stronger mechanism after changes. Adding, modifying (eg label text) or swapping some elements (eg `deck::select`) causes various overlapping / sizing issues. I noticed that resizing the window fixes the issue, so perhaps the resize does some extra work?

**Comment by Xeverous** [comment]
Obviously I would want to make an MCVE but currently working on other things.

**Comment by djowel** [comment]
> Obviously I would want to make an MCVE but currently working on other things.

Please do. I think a re-layout is needed when the layout changes (`view::layout()`), but like refresh, it has to be deferred. I was thinking of an API for this and I've been waiting for exactly this scenario to play out. I think I know what needs to be done.

**Comment by djowel** [comment]
> > Do you know any inconvenient about avoiding threading problems by this method?
> 
> This method is perfectly thread safe. ASIO takes care of the threading issues. View invalidation is guaranteed to happen only in the main thread.

@jpcima , just to be clear, you can call `view.refresh` any time, even in another thread. `view.refresh` is a deferred call and the actual OS refresh/invalidation will always happen in the main thread. ASIO takes care of the threading issues.

**Comment by Xeverous** [comment]
> I think a re-layout is needed when the layout changes

Check the MCVE in #143. This is a subset of the hierarchy I have in my project. It has a different symptom (unclickable buttons instead of sizing issues) but might come from the same problem.

**Comment by djowel** [comment]
Try doing a view.layout() before view.refresh()

**Comment by Xeverous** [comment]
I have updated elements, rebased my project, `view.layout()` solves various layouting issues, but I'm still getting blinking problem. Sometimes a fully solid color frame is being output which causes UI => black => UI frames which causes very unwanted blinking effect. This should never happen - I guess it is either a bug within cairo or elements on order/missing some renderong operations (probably elements since it seems to be tied to specific widgets).

**Comment by Xeverous** [comment]
Layout example also has this blinking problem.

**Comment by djowel** [comment]
Yes, it is a known issue. I haven't had time to look at it yet, and what's causing it. I'd love to get some help if you want to do some investigation.

**Comment by djowel** [comment]
> Layout example also has this blinking problem.

BTW, MacOS does not have this blinking issue on Layout. This might be a Windows thing, I'm not sure...

**Comment by djowel** [comment]
> UI => black => UI

Oh and what I am getting is UI => white => UI, when a menu item is selected. White, not black.

**Comment by redtide** [comment]
Just a guess, so it might be unrelated but most of the time I heard about ui blinking was a double buffering issue.
I just tried the layout example on Linux and the blinking happens only with the flow elements example.

**Comment by djowel** [comment]
> Just a guess, so it might be unrelated but most of the time I heard about ui blinking was a double buffering issue.
> I just tried the layout example on Linux and the blinking happens only with the flow elements example.

Yes, double-buffering is a typical issue. Does it blink when you select the menu or when you resize the window, or both?

**Comment by redtide** [comment]
Just tried the menus example and I get no white blinking on it, neither on the combobox on the layout one.
What I get is the black one when resizing with the flow elements.

**Comment by redtide** [comment]
Oh no wait sorry, IDK how I haven't noticed it but yes, is a black one also when selecting a menu item, on the layout one only.

**Comment by djowel** [comment]
> Oh no wait sorry, IDK how I haven't noticed it but yes, is a black one also when selecting a menu item, on the layout one only.

Haha! I think I got the insidious bug! :-)

**Comment by djowel** [comment]
Pushed fix to develop. The realize now bug did not manifest on MacOS because it has a different update mechanism with its Window server scheme.

**Comment by Xeverous** [comment]
Problem is gone.
